packages/shared_preferences/lib/shared_preferences.dart

```diff
@@ -38,6 +37,34 @@ class SharedPreferences {
     return _completer.future;
   }
 
+  static SharedPreferencesPlatform _platform;
+
+  /// Sets a custom [SharedPreferencesPlatform].
+  ///
+  /// This property can be set to use a custom platform implementation.
+  static set platform(SharedPreferencesPlatform platform) {
+    assert(_platform == null);
+    _platform = platform;
+  }
+
+  /// The SharedPreferences platform that's used by the plugin.
+  ///
+  /// The default value is [MethodChannelSharedPreferences] on Android or iOS.
+  static SharedPreferencesPlatform get platform {
+    if (_platform == null) {
+      switch (defaultTargetPlatform) {
+        case TargetPlatform.android:
+        case TargetPlatform.iOS:
+          _platform = MethodChannelSharedPreferences();
+          break;
+        default:
+          throw UnsupportedError(
```

Comment:
It's worth discussing what the default should be for new projects, but for existing plugins I strongly recommend we default to the old behavior, which is the method channel implementation. This change would break at least two unofficial federated implementations for people who set their desktop projects' TargetPlatform to Fuchsia (which FDE's examples use a sort of "neutral" default).

Reply:
(And in general, anything whitelist-style should be avoided in plugins going forward IMO, since it works against federation. I had issues with path_provider's Dart code allowing some code paths only for iOS, for instance, when in fact they would apply to macOS as well.)

Reply:
I like your ideas but I'm not sure what you're recommending about the code. What do you think the implementation should look like?

@amirh This conversation will be relevant to you, too, as this is the basic architecture that you proposed for the Battery plugin.

Reply:
As an existing plugin, I think the whole implementation of this function should just be:
```
if (_platform == null) {
  MethodChannelSharedPreferences();
}
return _platform;
```

That preserves behavior for all unofficial implementations, and requires implementations that want a new and different behavior to opt into it.

Reply:
updated to use MethodChannelSharedPreferences if no other platform-specific instance is provided.

Reply:
Stuart's point makes sense to me for existing plugins, the only drawback is that in cases where there's no platform implementation the error message is less clear but I'm willing to live with that if we think we're going to break folks otherwise.

Ideally we'd want existing platform implementations to migrate to the new approach, so maybe worth considering logging an error when we're falling back to a method channel implementation.